src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/FallbackProjectManager.cs

```diff
@@ -0,0 +1,163 @@
+        {
+            // If this is a fallback project, then Roslyn may not track documents in the project, so these dynamic file notifications
+            // are the only way to know about files in the project.
+            RemoveFallbackDocument(projectId, filePath, projectFilePath);
+        }
+    }
+
+    private void AddFallbackProject(ProjectId projectId, string filePath)
+    {
+        var project = TryFindProjectForProjectId(projectId);
+        if (project?.FilePath is null)
+        {
+            return;
+        }
+
+        var intermediateOutputPath = Path.GetDirectoryName(project.CompilationOutputInfo.AssemblyPath);
+        if (intermediateOutputPath is null)
+        {
+            return;
+        }
+
+        if (!ImmutableInterlocked.Update(ref _fallbackProjectIds, (set, id) => set.Add(id), project.Id))
+        {
+            return;
+        }
+
+        var rootNamespace = project.DefaultNamespace ?? "ASP";
+        var hostProject = new HostProject(project.FilePath, intermediateOutputPath, FallbackRazorConfiguration.Latest, rootNamespace, project.Name);
+
+        _projectSnapshotManagerAccessor.Instance.ProjectAdded(hostProject);
+
+        AddFallbackDocument(hostProject.Key, filePath, project.FilePath);
+
+        var configurationFilePath = Path.Combine(intermediateOutputPath, _languageServerFeatureOptions.ProjectConfigurationFileName);
+
+        _projectConfigurationFilePathStore.Set(hostProject.Key, configurationFilePath);
+    }
+
+    private void AddFallbackDocument(ProjectKey projectKey, string filePath, string projectFilePath)
+    {
+        var hostDocument = CreateHostDocument(filePath, projectFilePath);
+        var textLoader = new FileTextLoader(filePath, defaultEncoding: null);
+        _projectSnapshotManagerAccessor.Instance.DocumentAdded(projectKey, hostDocument, textLoader);
+    }
+
+    private static HostDocument CreateHostDocument(string filePath, string projectFilePath)
+    {
+        var targetPath = filePath.StartsWith(projectFilePath, FilePathComparison.Instance)
+            ? filePath[projectFilePath.Length..]
+            : filePath;
+        var hostDocument = new HostDocument(filePath, targetPath);
+        return hostDocument;
+    }
+
+    private void RemoveFallbackDocument(ProjectId projectId, string filePath, string projectFilePath)
+    {
+        var project = TryFindProjectForProjectId(projectId);
+        if (project is null)
+        {
+            return;
+        }
+
+        var projectKey = ProjectKey.From(project);
+        if (projectKey is not { } razorProjectKey)
+        {
+            return;
+        }
+
+        var hostDocument = CreateHostDocument(filePath, projectFilePath);
+        _projectSnapshotManagerAccessor.Instance.DocumentRemoved(razorProjectKey, hostDocument);
+    }
+
+    private Project? TryFindProjectForProjectId(ProjectId projectId)
+    {
+        if (_projectSnapshotManagerAccessor.Instance.Workspace is not { } workspace)
```

Comment:
Can this ever actually be null?

Reply:
Probably not, aside from some weird MEF horror story